fclient/performjoin.go

```diff
@@ -0,0 +1,238 @@
+package fclient
+
+import (
+	"context"
+	"crypto/ed25519"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/matrix-org/gomatrixserverlib/spec"
+	"github.com/sirupsen/logrus"
+)
+
+type PerformJoinInput struct {
+	UserID     *spec.UserID
+	RoomID     string
+	ServerName spec.ServerName
+	Content    map[string]interface{}
+	Unsigned   map[string]interface{}
+
+	PrivateKey ed25519.PrivateKey
+	KeyID      gomatrixserverlib.KeyID
+	KeyRing    *gomatrixserverlib.KeyRing
+
+	EventProvider func(
+		ctx context.Context, fedClient FederationClient,
+		keyRing gomatrixserverlib.JSONVerifier, origin, server spec.ServerName,
+	) gomatrixserverlib.EventProvider
+}
+
+type PerformJoinCallbacks struct {
+	FederationFailure func(serverName spec.ServerName)
+	FederationSuccess func(serverName spec.ServerName)
+}
+
+func PerformJoin(
+	ctx context.Context,
+	fedClient FederationClient,
+	input PerformJoinInput,
+	callbacks PerformJoinCallbacks,
+) (*gomatrixserverlib.HeaderedEvent, gomatrixserverlib.StateResponse, error) {
+	origin := input.UserID.Domain()
+
+	// Try to perform a make_join using the information supplied in the
+	// request.
+	respMakeJoin, err := fedClient.MakeJoin(
+		ctx,
+		origin,
+		input.ServerName,
+		input.RoomID,
+		input.UserID.Raw(),
+	)
+	if err != nil {
+		// TODO: Check if the user was not allowed to join the room.
+		callbacks.FederationFailure(input.ServerName)
+		return nil, nil, fmt.Errorf("r.federation.MakeJoin: %w", err)
```

Comment:
The use of callbacks here feels weird. I'd expect us to have a defined errror type e.g `FederationFailureError` which we then return, which callers can then type-switch against.